File: ios/ReactNativeCharts/combine/BFCombinedChartRenderer.swift
//
//  BFCombinedChartRenderer.swift
//  RNCharts
//
//  Created by Taylor Johnson on 6/17/20.
//

import Foundation
import Charts

// Support Highlighting Candle and Line charts where the horizontal highlight
// line follows the Highlight's y value
class BFCombinedChartRenderer: CombinedChartRenderer {
	
	@objc public override init(chart: CombinedChartView, animator: Animator, viewPortHandler: ViewPortHandler)
     {
        super.init(chart: chart, animator: animator, viewPortHandler: viewPortHandler)

         self.chart = chart

         createRenderers()
     }

     /// Creates the renderers needed for this combined-renderer in the required order. Also takes the DrawOrder into consideration.
     internal func createRenderers()
     {
         subRenderers = [DataRenderer]()

         guard let chart = chart else { return }

         for order in drawOrder
         {
             switch (order)
             {
             case .bar:
                 if chart.barData !== nil
                 {
                     subRenderers.append(BarChartRenderer(dataProvider: chart, animator: animator, viewPortHandler: viewPortHandler))
                 }
                 break

             case .line:
                 if chart.lineData !== nil
                 {
                     subRenderers.append(BFLineChartRenderer(dataProvider: chart, animator: animator, viewPortHandler: viewPortHandler))
                 }
                 break

             case .candle:
                 if chart.candleData !== nil
                 {
                     subRenderers.append(BFCandleStickChartRenderer(dataProvider: chart, animator: animator, viewPortHandler: viewPortHandler))
                 }
                 break

             case .scatter:
                 if chart.scatterData !== nil
                 {
                     subRenderers.append(ScatterChartRenderer(dataProvider: chart, animator: animator, viewPortHandler: viewPortHandler))
                 }
                 break

             case .bubble:
                 if chart.bubbleData !== nil
                 {
                     subRenderers.append(BubbleChartRenderer(dataProvider: chart, animator: animator, viewPortHandler: viewPortHandler))
                 }
                 break
             }
         }

     }

	open override func drawHighlighted(context: CGContext, indices: [Highlight])
	{
        for renderer in subRenderers
        {
            var data: ChartData?

            if renderer is BarChartRenderer
            {
                data = (renderer as! BarChartRenderer).dataProvider?.barData
            }
            else if renderer is LineChartRenderer
            {
                data = (renderer as! BFLineChartRenderer).dataProvider?.lineData
            }
            else if renderer is CandleStickChartRenderer
            {
                data = (renderer as! BFCandleStickChartRenderer).dataProvider?.candleData
            }
            else if renderer is ScatterChartRenderer
            {
                data = (renderer as! ScatterChartRenderer).dataProvider?.scatterData
            }
            else if renderer is BubbleChartRenderer
            {
                data = (renderer as! BubbleChartRenderer).dataProvider?.bubbleData
            }

            let dataIndex: Int? = {
                guard let data = data else { return nil }
                return (chart?.data as? CombinedChartData)?
                        .allData
                        .firstIndex(of: data)
            }()

            let dataIndices = indices.filter{ $0.dataIndex == dataIndex || $0.dataIndex == -1 }

            renderer.drawHighlighted(context: context, indices: dataIndices)
        }
	}


	class BFCandleStickChartRenderer: CandleStickChartRenderer {

		/// Checks if the provided entry object is in bounds for drawing considering the current animation phase.
		internal func isInBoundsX(entry e: ChartDataEntry, dataSet: IBarLineScatterCandleBubbleChartDataSet) -> Bool
		{
            let entryIndex = dataSet.entryIndex(entry: e)
            return Double(entryIndex) < Double(dataSet.entryCount) * animator.phaseX
		}

		open override func drawHighlighted(context: CGContext, indices: [Highlight])
		{
            guard
                let dataProvider = dataProvider,
                let candleData = dataProvider.candleData
                else { return }

            context.saveGState()

            for high in indices
            {
                    guard
                        let set = candleData.getDataSetByIndex(high.dataSetIndex) as? ICandleChartDataSet,
                        set.isHighlightEnabled
                        else { continue }

                    guard let e = set.entryForXValue(high.x, closestToY: high.y) as? CandleChartDataEntry else { continue }

                    if !isInBoundsX(entry: e, dataSet: set)
                    {
                        continue
                    }

                    let trans = dataProvider.getTransformer(forAxis: set.axisDependency)

                    context.setStrokeColor(set.highlightColor.cgColor)
                    context.setLineWidth(set.highlightLineWidth)

                    if set.highlightLineDashLengths != nil
                    {
                        context.setLineDash(phase: set.highlightLineDashPhase, lengths: set.highlightLineDashLengths!)
                    }
                    else
                    {
                        context.setLineDash(phase: 0.0, lengths: [])
                    }

                    let pt = trans.pixelForValues(x: e.x, y: high.y)

                    high.setDraw(pt: pt)

                    // draw the lines
                    drawHighlightLines(context: context, point: pt, set: set)
            }

            context.restoreGState()
		}
	}


	class BFLineChartRenderer: LineChartRenderer {

		/// Checks if the provided entry object is in bounds for drawing considering the current animation phase.
		internal func isInBoundsX(entry e: ChartDataEntry, dataSet: IBarLineScatterCandleBubbleChartDataSet) -> Bool
		{
            let entryIndex = dataSet.entryIndex(entry: e)
            return Double(entryIndex) < Double(dataSet.entryCount) * animator.phaseX
		}

		open override func drawHighlighted(context: CGContext, indices: [Highlight])
		{
            guard
                let dataProvider = dataProvider,
                let lineData = dataProvider.lineData
                else { return }

            let chartXMax = dataProvider.chartXMax

            context.saveGState()

            for high in indices
            {
                    guard let set = lineData.getDataSetByIndex(high.dataSetIndex) as? ILineChartDataSet
                            , set.isHighlightEnabled
                            else { continue }

                    guard let e = set.entryForXValue(high.x, closestToY: high.y) else { continue }

                    if !isInBoundsX(entry: e, dataSet: set)
                    {
                            continue
                    }

                    context.setStrokeColor(set.highlightColor.cgColor)
                    context.setLineWidth(set.highlightLineWidth)
                    if set.highlightLineDashLengths != nil
                    {
                            context.setLineDash(phase: set.highlightLineDashPhase, lengths: set.highlightLineDashLengths!)
                    }
                    else
                    {
                            context.setLineDash(phase: 0.0, lengths: [])
                    }

                    let x = e.x // get the x-position
                    let y = high.y

                    if x > chartXMax * animator.phaseX
                    {
                            continue
                    }

                    let trans = dataProvider.getTransformer(forAxis: set.axisDependency)

                    let pt = trans.pixelForValues(x: x, y: y)

                    high.setDraw(pt: pt)

                    // draw the lines
                    drawHighlightLines(context: context, point: pt, set: set)
            }

            context.restoreGState()
		}
	}

}

File: android/src/main/java/com/github/wuxudong/rncharts/charts/BFCombinedChartRenderer.java
package com.github.wuxudong.rncharts.charts;

import android.graphics.Canvas;
import com.github.mikephil.charting.animation.ChartAnimator;
import com.github.mikephil.charting.charts.Chart;
import com.github.mikephil.charting.charts.CombinedChart;
import com.github.mikephil.charting.data.CandleData;
import com.github.mikephil.charting.data.CandleEntry;
import com.github.mikephil.charting.data.ChartData;
import com.github.mikephil.charting.data.CombinedData;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.highlight.Highlight;
import com.github.mikephil.charting.interfaces.dataprovider.*;
import com.github.mikephil.charting.interfaces.datasets.ICandleDataSet;
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet;
import com.github.mikephil.charting.renderer.*;
import com.github.mikephil.charting.utils.MPPointD;
import com.github.mikephil.charting.utils.ViewPortHandler;

public class BFCombinedChartRenderer extends CombinedChartRenderer {
    public BFCombinedChartRenderer(CombinedChart chart, ChartAnimator animator, ViewPortHandler viewPortHandler) {
        super(chart, animator, viewPortHandler);
    }

    /**
     * Creates the renderers needed for this combined-renderer in the required order. Also takes the DrawOrder into
     * consideration.
     */
    @Override
    public void createRenderers() {

        mRenderers.clear();

        CombinedChart chart = (CombinedChart)mChart.get();
        if (chart == null)
            return;

        CombinedChart.DrawOrder[] orders = chart.getDrawOrder();

        for (CombinedChart.DrawOrder order : orders) {

            switch (order) {
                case BAR:
                    if (chart.getBarData() != null)
                        mRenderers.add(new BFBarChartRenderer(chart, mAnimator, mViewPortHandler));
                    break;
                case BUBBLE:
                    if (chart.getBubbleData() != null)
                        mRenderers.add(new BFBubbleChartRenderer(chart, mAnimator, mViewPortHandler));
                    break;
                case LINE:
                    if (chart.getLineData() != null)
                        mRenderers.add(new BFLineChartRenderer(chart, mAnimator, mViewPortHandler));
                    break;
                case CANDLE:
                    if (chart.getCandleData() != null)
                        mRenderers.add(new BFCandleStickChartRenderer(chart, mAnimator, mViewPortHandler));
                    break;
                case SCATTER:
                    if (chart.getScatterData() != null)
                        mRenderers.add(new BFScatterChartRenderer(chart, mAnimator, mViewPortHandler));
                    break;
            }
        }
    }

    @Override
    public void drawHighlighted(Canvas c, Highlight[] indices) {

        Chart chart = mChart.get();
        if (chart == null) return;

        for (DataRenderer renderer : mRenderers) {
            ChartData data = null;

            if (renderer instanceof BFBarChartRenderer)
                data = ((BFBarChartRenderer)renderer).getChartDataProvider().getBarData();
            else if (renderer instanceof BFLineChartRenderer)
                data = ((BFLineChartRenderer)renderer).getChartDataProvider().getLineData();
            else if (renderer instanceof BFCandleStickChartRenderer)
                data = ((BFCandleStickChartRenderer)renderer).getChartDataProvider().getCandleData();
            else if (renderer instanceof BFScatterChartRenderer)
                data = ((BFScatterChartRenderer)renderer).getChartDataProvider().getScatterData();
            else if (renderer instanceof BFBubbleChartRenderer)
                data = ((BFBubbleChartRenderer)renderer).getChartDataProvider().getBubbleData();

            int dataIndex = data == null ? -1
                    : ((CombinedData)chart.getData()).getAllData().indexOf(data);

            mHighlightBuffer.clear();

            for (Highlight h : indices) {
                if (h.getDataIndex() == dataIndex || h.getDataIndex() == -1)
                    mHighlightBuffer.add(h);
            }

            renderer.drawHighlighted(c, mHighlightBuffer.toArray(new Highlight[mHighlightBuffer.size()]));
        }
    }

    static class BFCandleStickChartRenderer extends CandleStickChartRenderer {
        public BFCandleStickChartRenderer(CandleDataProvider chart, ChartAnimator animator, ViewPortHandler viewPortHandler) {
            super(chart, animator, viewPortHandler);
        }

        protected CandleDataProvider getChartDataProvider() {
            return mChart;
        }

        @Override
        public void drawHighlighted(Canvas c, Highlight[] indices) {

            CandleData candleData = mChart.getCandleData();

            for (Highlight high : indices) {

                ICandleDataSet set = candleData.getDataSetByIndex(high.getDataSetIndex());

                if (set == null || !set.isHighlightEnabled())
                    continue;

                CandleEntry e = set.getEntryForXValue(high.getX(), high.getY());

                if (!isInBoundsX(e, set))
                    continue;

                MPPointD pix = mChart.getTransformer(set.getAxisDependency()).getPixelForValues(e.getX(), high.getY());

                high.setDraw((float) pix.x, (float) pix.y);

                // draw the lines
                drawHighlightLines(c, (float) pix.x, (float) pix.y, set);
            }
        }
    }

    static class BFBarChartRenderer extends BarChartRenderer {
        public BFBarChartRenderer(BarDataProvider chart, ChartAnimator animator, ViewPortHandler viewPortHandler) {
            super(chart, animator, viewPortHandler);
        }

        protected BarDataProvider getChartDataProvider() {
            return mChart;
        }
    }

    static class BFLineChartRenderer extends LineChartRenderer {
        public BFLineChartRenderer(LineDataProvider chart, ChartAnimator animator, ViewPortHandler viewPortHandler) {
            super(chart, animator, viewPortHandler);
        }

        protected LineDataProvider getChartDataProvider() {
            return mChart;
        }

        @Override
        public void drawHighlighted(Canvas c, Highlight[] indices) {

            LineData lineData = mChart.getLineData();

            for (Highlight high : indices) {

                ILineDataSet set = lineData.getDataSetByIndex(high.getDataSetIndex());

                if (set == null || !set.isHighlightEnabled())
                    continue;

                Entry e = set.getEntryForXValue(high.getX(), high.getY());

                if (!isInBoundsX(e, set))
                    continue;

                MPPointD pix = mChart.getTransformer(set.getAxisDependency()).getPixelForValues(e.getX(), high.getY());

                high.setDraw((float) pix.x, (float) pix.y);

                // draw the lines
                drawHighlightLines(c, (float) pix.x, (float) pix.y, set);
            }
        }
    }

    static class BFScatterChartRenderer extends ScatterChartRenderer {
        public BFScatterChartRenderer(ScatterDataProvider chart, ChartAnimator animator, ViewPortHandler viewPortHandler) {
            super(chart, animator, viewPortHandler);
        }

        protected ScatterDataProvider getChartDataProvider() {
            return mChart;
        }
    }

    static class BFBubbleChartRenderer extends BubbleChartRenderer {
        public BFBubbleChartRenderer(BubbleDataProvider chart, ChartAnimator animator, ViewPortHandler viewPortHandler) {
            super(chart, animator, viewPortHandler);
        }

        protected BubbleDataProvider getChartDataProvider() {
            return mChart;
        }
    }
}
